presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ActualProperties.java

```diff
@@ -169,22 +171,43 @@ public boolean isStreamRepartitionEffective(Collection<VariableReferenceExpressi
         return global.isStreamRepartitionEffective(keys, constants.keySet());
     }
 
-    public ActualProperties translate(Function<VariableReferenceExpression, Optional<VariableReferenceExpression>> translator)
+    public ActualProperties translate(Function<RowExpression, Optional<RowExpression>> translator)
     {
         Map<VariableReferenceExpression, ConstantExpression> translatedConstants = new HashMap<>();
         for (Map.Entry<VariableReferenceExpression, ConstantExpression> entry : constants.entrySet()) {
-            Optional<VariableReferenceExpression> translatedKey = translator.apply(entry.getKey());
-            if (translatedKey.isPresent()) {
-                translatedConstants.put(translatedKey.get(), entry.getValue());
-            }
+            Optional<RowExpression> translatedKey = translator.apply(entry.getKey());
+            checkState(!translatedKey.isPresent() || translatedKey.get() instanceof VariableReferenceExpression, "Expect translatedKey to be VariableReferenceExpression");
```

Comment:
This is using the passed in translator (which translates a `VariableReferenceExpression` to a `Optional<VariableReferenceExpression>`) to translate the keys on `constants`. So the translated result should be `Optional<VariableReferenceExpression>`.

Reply:
@rongrong : I see. So does it mean now constant will always binding to some variables and get referred by a `VariableReferenceExpression` ? (this is slightly different from what we did before right,? since it directly refer to a `NullableValue` in `ArgumentBinding`) .

This doesn't change the semantic / correctness of the code. Just try to double check my understanding ;)

Reply:
Please add the actual type, just in case it ever fails

Reply:
Actually you can simply remove this check. The `(VariableReferenceExpression) key` below will do the job.

Reply:
I added the `checkState` so if this ever fails the error message is clearer. I can remove it as well, since the cast is just next line. What do you prefer? @arhimondr